apps/www/_blog/2023-01-25-type-constraints-in-50-lines-of-sql.mdx

```diff
@@ -0,0 +1,271 @@
+---
+title: Type Constraints in 50 lines of SQL
+description: Creating validated data types in Postgres
+author: oli_rice
+image:
+thumb:
+tags:
+  - postgres
+  - planetpg
+date: '2023-01-25'
+toc_depth: 3
+---
+
+PostgreSQL has a rich and extensible type system. Beyond defining custom enums and composite types, we can:
+
+- apply data validation rules
+- override comparison operators like `=` / `+` / `-`
+- create custom aggregations
+- define casting rules between types
+
+until our types are indistinguishable from built-ins. In this article we’ll focus on validation and ergonomics, and touch on a few other concepts.
+
+To illustrate, we’ll create an `semver` data type to represent [Semantic Versioning](https://semver.org) values. We’ll then add validation rules to make invalid states unrepresentable.
+
+## SemVer
+
+A (very) loose primer on SemVer:
+
+SemVer is a specification for representing software versions that communicate information about backwards compatibility. The type is typically represented as a string with 5 components.
+
+<IMAGE HERE>
```

Comment:
TODO: image here that is comparable to
<img width="386" alt="Screenshot 2023-01-25 at 8 34 35 AM" src="https://user-images.githubusercontent.com/12958657/214591070-de67c30a-d224-47fa-ae26-6c7bf1a33c07.png">